src/ImageSharp/Common/ParallelUtils/ParallelHelper.cs

```diff
@@ -0,0 +1,151 @@
+                    });
+        }
+
+        /// <summary>
+        /// Iterate through the rows of a rectangle in optimized batches defined by <see cref="RowInterval"/>-s
+        /// instantiating a temporary buffer for each <paramref name="body"/> invocation.
+        /// </summary>
+        public static void IterateRowsWithTempBuffer<T>(
+            Rectangle rectangle,
+            in ParallelExecutionSettings parallelSettings,
+            Action<RowInterval, Memory<T>> body)
+            where T : struct
+        {
+            int maxSteps = DivideCeil(rectangle.Width * rectangle.Height, parallelSettings.MinimumPixelsProcessedPerTask);
+
+            int numOfSteps = Math.Min(parallelSettings.MaxDegreeOfParallelism, maxSteps);
+
+            MemoryAllocator memoryAllocator = parallelSettings.MemoryAllocator;
+
+            // Avoid TPL overhead in this trivial case:
+            if (numOfSteps == 1)
+            {
+                var rows = new RowInterval(rectangle.Top, rectangle.Bottom);
+                using (IMemoryOwner<T> buffer = memoryAllocator.Allocate<T>(rectangle.Width))
+                {
+                    body(rows, buffer.Memory);
+                }
+
+                return;
+            }
+
+            int verticalStep = DivideCeil(rectangle.Height, numOfSteps);
+
+            var parallelOptions = new ParallelOptions() { MaxDegreeOfParallelism = numOfSteps };
+
+            Parallel.For(
+                0,
+                numOfSteps,
+                parallelOptions,
+                i =>
+                    {
+                        int yMin = rectangle.Top + (i * verticalStep);
+                        int yMax = Math.Min(yMin + verticalStep, rectangle.Bottom);
+
+                        var rows = new RowInterval(yMin, yMax);
+
+                        using (IMemoryOwner<T> buffer = memoryAllocator.Allocate<T>(rectangle.Width))
+                        {
+                            body(rows, buffer.Memory);
+                        }
+                    });
+        }
+
+        /// <summary>
+        /// Iterate through the rows of a rectangle in optimized batches defined by <see cref="RowInterval"/>-s
+        /// instantiating a temporary buffer for each <paramref name="body"/> invocation.
+        /// </summary>
+        public static void IterateRowsWithTempBuffer<T>(
+            Rectangle rectangle,
+            Configuration configuration,
+            Action<RowInterval, Memory<T>> body)
+            where T : struct
+        {
+            IterateRowsWithTempBuffer(rectangle, configuration.GetParallelSettings(), body);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DivideCeil(int dividend, int divisor)
+        {
+            // TODO: Is there a more efficient way to calculate this?
+            int result = dividend / divisor;
+            return dividend % divisor == 0 ? result : result + 1;
+        }
```

Comment:
`(dividend > 0) ? 1 + ((dividend - 1) / divisor) : (dividend / divisor)` Maybe? Seems to work in my local tests.